core-bundle/src/Repository/AltchaRepository.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Repository;
+
+use Contao\CoreBundle\Entity\Altcha;
+use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Types\Types;
+use Symfony\Bridge\Doctrine\ManagerRegistry;
+
+/**
+ * @template-extends ServiceEntityRepository<Altcha>
+ *
+ * @internal
+ */
+class AltchaRepository extends ServiceEntityRepository
+{
+    private readonly Connection $connection;
+
+    public function __construct(ManagerRegistry $registry)
+    {
+        parent::__construct($registry, Altcha::class);
+
+        /** @var Connection $connection */
+        $connection = $registry->getConnection();
+
+        $this->connection = $connection;
+    }
+
+    public function purgeExpiredChallenges(): void
+    {
+        $table = $this->getClassMetadata()->getTableName();
+
+        $this->connection->executeStatement(
+            "DELETE FROM $table WHERE expires < :expires",
+            ['expires' => new \DateTimeImmutable()],
+            ['expires' => Types::DATETIME_IMMUTABLE],
+        );
+    }
+
+    public function isReplay(string $challenge): bool
+    {
+        $table = $this->getClassMetadata()->getTableName();
```

Comment:
Haven't seen this style of writing queries in Doctrine repositories anywhere. I think we should use the query builder using `$this->createQueryBuilder()` - that'd be more Doctrine-style.

Reply:
We use this kind of query in the core:

https://github.com/contao/contao/blob/5f54aa767f249fc23a6325bfff2dd00b4b004a22/core-bundle/src/EventListener/DataContainer/TemplateOptionsListener.php#L174-L178

https://github.com/contao/contao/blob/5f54aa767f249fc23a6325bfff2dd00b4b004a22/core-bundle/src/Migration/Version500/OrderFieldMigration.php#L104-L115

I wouldn‘t say that it is less Doctrine-style than using the query builder.